### DISCUSSION TITLE
"Hybird" Mode

### DISCUSSION THREAD

**Discussion by Stoppedpuma** [YaLTeR/niri#873] [Ideas]
I use niri as my default window manager on my laptop and find the experience to be great, so much in fact that I constantly find myself wanting to bring this experience to my desktop computer. I find myself running into one large wall though on desktop, the way of handling scrollable workspaces does not work that well for a desktop experience. On portable devices such as my laptop I find this great, but on desktop not so much and I always default back to more traditional tiling window managers such as Sway. I have spent a bit of time thinking and have thought of a way which merges the feel of a scrollable window manager such as niri and a traditional one such as Sway, an in-between mode hence the title "hybrid" (I also just couldn't think of a better name / description). This mode works by scrolling when the edge is reached manually instead of when new windows are created. The strip behaviour is kept, but managed into individual workspace like groups which I'm calling "Hybrid Workspaces" because I can't think of a better name, switching actual workspaces should still work as intended and create a new strip. I think this could also open the door to more benefits such as dynamic tiling options. 

Below are some designs of possible implementations.

"Classic Scrolling": This mimics the current way that niri shows windows on the left and right sides

![image](https://github.com/user-attachments/assets/0d2cb7d2-381c-4214-af96-de8ad48fae1a)

"Arrows": Arrows has little arrows on the sides showing how many "Hybird Workspaces" are to the left / right. The number of arrows should follow how many times the user can scroll in that direction.

![Group 9](https://github.com/user-attachments/assets/f1cdfc68-cacf-442d-a854-75b90582097a)

![image](https://github.com/user-attachments/assets/e3402b58-b282-4811-bd69-c9ad5c9cd71e)

"Dot indicator": The dot indicator is a classic tray approach which shows where the user is currently located.

![Group 10](https://github.com/user-attachments/assets/e8c9edd0-1da4-405b-a138-b48e558876f6)

![image](https://github.com/user-attachments/assets/cffe2a69-831b-4583-abb0-d3a5b1348b50)

**Comment by YaLTeR** [comment]
Hm, so more or less just have two dimensions of workspaces? Where every strip contains monitor-sized workspaces that you can scroll left and right, and up and down scrolls through strips. And then every workspace is a regular tiling WM sort of workspace.

This concept sounds like it would be easier to try as a Hyprland plugin or something, I feel like implementing "regular tiling" fixed to monitor size would be more effort in niri.

**Reply by Stoppedpuma** [reply]
> Hm, so more or less just have two dimensions of workspaces? Where every strip contains monitor-sized workspaces that you can scroll left and right, and up and down scrolls through strips. And then every workspace is a regular tiling WM sort of workspace.

Yes, this is the main thing preventing me from converting my desktop completely towards niri besides things like floating windows and scratch pads but those seems to already be in the works!

> This concept sounds like it would be easier to try as a Hyprland plugin or something, I feel like implementing "regular tiling" fixed to monitor size would be more effort in niri.

I have considered this as well, the problem I have here is finding where would be best to create a solution if I was going to. I personally would end up avoiding Hyprland as it tends to break tons of things and have half baked implementations quite often. Others such as Sway don't have plugin support either which makes this a bit harder. On the other hand, niri seems to have a solid foundation for this already with it being a scrolling window manager which was why I raised this here before attempting my own solution. I think this opens a bunch of possible benefits to niri such as dynamic tiling or finer grained window grouping which could be used outside of this feature as well.

On the topic of finer grained window grouping, maybe this feature would work better as "window containers" which can be resized so they can be used side by side with how the current behaviour works? Not sure if this would make it easier to implement or not, especially since a by default option would be appreciated.

**Reply by YaLTeR** [reply]
> On the topic of finer grained window grouping, maybe this feature would work better as "window containers" which can be resized so they can be used side by side with how the current behaviour works? Not sure if this would make it easier to implement or not, especially since a by default option would be appreciated.

Maybe. I think this would fit more naturally in the code as a new layer of hierarchy on top of columns. Though, my main question is how the UX around it should work in a way that is not confusing (un/grouping columns, moving windows in and out, etc.).

Also there's this synchronized horizontal resize idea that may cover some of the uses of the grouped windows idea: https://github.com/YaLTeR/niri/discussions/868

**Reply by Stoppedpuma** [reply]
> Maybe. I think this would fit more naturally in the code as a new layer of hierarchy on top of columns. Though, my main question is how the UX around it should work in a way that is not confusing (un/grouping columns, moving windows in and out, etc.).

I think we could borrow a few ideas from sway here with coloured borders depending on the group type. Adding and removing could be done by selecting multiple windows similarly to shift clicking to select multiple and then using a keybind to create a group type (e.g regular group, tabbed group, vertical tabbed group, etc). Basically just colours to coordinate group type. Things could be added or removed to a group by dragging them in and out, possibly a keybind as well? Accordingly, it should also be easy to resize a group, a keybind option would be useful here as well, something like mod+alt+right and the windows should scale along with group resizes.

Below is a basic, very rushed design showing this. I haven't had much time to give this proper thinking UI and UX wise so I'll revise this when I have more time over the next few days.

![image](https://github.com/user-attachments/assets/504ad91c-dcd1-4777-aacf-a2eb5a64d6f2)

**Reply by Stoppedpuma** [reply]
I have revised the designs mentioned above and rethought UX in particular about how specifically tabbed groups should work as I saw another discussion about tabbed groups here.

Hyprland requires an additional key to be used for switching between the windows in a group, for this example, lets use arrow keys and alt. Say the user is moving to the right of the strip, they wouldn't want to have to scroll through every window in a tabbed container, so instead we should hold alt when the container is selected to be able to move through the windows, if it isn't held, the users next right step goes past the container.

Revised design for horizontal tab container here:

![Group 15](https://github.com/user-attachments/assets/38a3ab6e-fe35-40cf-afd4-9b9acce8e946)

As there's a bit of open space towards the sides with this design style, window decorations can be removed by default and buttons like close can be put onto the open area such as the top right.

Regarding vertical tab containers, it's pretty much the same thing. I've decided to put the vertical windows on the side instead of actually vertical like what you would find in sway or i3 because as niri is scrolling, chances are, the user wants to use more horizontal space and reserve vertical space for the application. The same idea of moving buttons like close and hiding window decorations for windows in these groups can be applied here such as above the tabs in the top right.

Vertical tabs design:

![Group 17 (1)](https://github.com/user-attachments/assets/fdcf3053-8139-41d7-b059-d7426e2de56f)

Some additional UX can be added here such as adding close buttons to the tabs which show on hover. This can be applied to both container types. Additionally, tabs should be able to be dragged around to be re-arranged. Dragging windows out of the group should work from both a tab level, and window level. From the window level though, a modifier should be held such as alt since it was mentioned above, otherwise moving with a bind such as mod+m1 should move the entire container, mod+alt+m1 in this case should allow moving the windows inside of tabbed containers.

As for "Regular Groups", I think these should have an option to be required to hold a key to switch through the windows inside, or just go through them by default as this one I can see being pretty 50/50 in preference.

**Reply by YaLTeR** [reply]
Thanks for the concepts and for the info on how it works in Hyprland.

Yeah, it's a bit unfortunate that adding this extra layer of grouping requires an extra set of binds. I feel like at this point it becomes too easy to tangle in the various layers of keybinds.

**Reply by Stoppedpuma** [reply]
This could always be done the opposite way where it goes through windows in a tab by default and a key could be used to ignore the windows in a group if you think that simplifies things? Otherwise I'm not too sure how to have this work efficiently for those using keys to switch between windows. Sway doesn't use a key at all and just goes through the windows, but as mentioned previously, this is pretty inefficient especially for the design of niri as it requires going through all the remaining windows in said direction before going to the next window outside of the container.

Of course a solution to this is to just not add horizontal tabbed groups which could work, but seem to be pretty 50/50 in preference compared to vertical tabbed groups. I think regular groups don't have to worry about the key behaviour as it's basically just the same behaviour you would find in a traditional window manager.

**Reply by mastoca** [reply]
@Stoppedpuma Could some of these ideas be tested by running niri in niri? I've customized my keybinds to a brittle degree but I could see separate niri instances could simulated 'grouped' like you describe if your display is massive enough to enjoy that. It will likely be very odd behavior but it would allow you to validate some these ideas.

I came from xmonad and had a host of tiled panels for different use cases, but I've enjoyed the smoothness of the endless scroll since I dedicated a separate trackpad to just do that with my left hand (contemplating using a keyboard with a knob to perhaps use that instead once it arrives).

In terms of ux, personally, I've found tabs to be a bit too much work to traverse and locate things.

**Reply by Stoppedpuma** [reply]
> Could some of these ideas be tested by running niri in niri?

Yes and no, keybinds will overlap and end up being excluded in the embedded niri instance(s), additionally the occasional app may refuse to open and shout about requiring a valid wayland display.

> In terms of ux, personally, I've found tabs to be a bit too much work to traverse and locate things.

These were added after the original intention of this discussion as I found another discussion with this at #876 and figured I may as well add it here since it's also pretty similar to what you would find in a more traditional window manager. The main things I'm after here is being able to make resizable window groups to be able to have something similar to horizontal and vertical scrolling workspaces, not so much the tabs part.

**Reply by thepragandsensdiary** [reply]
I'll say Hy to the bird too

**Comment by teohhanhui** [comment]
I'm trying out Niri a bit, and the infinite horizontal scrolling means I easily lose track of where I am, and trying to find back previous windows becomes a chore where I have to keep scrolling left and right to locate things.

Or in other words, I'm a user who doesn't want to manually arrange windows into workspaces.

I wish there's a way for Niri to automatically push windows into the next workspace past a certain horizontal threshold. That will make the overview actually useful for me, instead of just being a zoomed out view of still a single infinite horizontal scrolling workspace.

**Reply by thepragandsensdiary** [reply]
maybe a "max_workspace_windows" ??
a default value for new workspaces, and a different value to pre-determined ones???

or maybe just use what you actually need lol

**Reply by teohhanhui** [reply]
> max_workspace_windows

This is unlikely to work well. I think max workspace width (e.g. with `proportion` unit) makes much more sense...

> or maybe just use what you actually need lol

Can you enlighten me on what that would be? Because I don't believe there's currently such a window manager / desktop environment that works the way I described.

The closest thing is probably what's described as a concept by GNOME's designers.
